### PR TITLE
[SiVal] I2C test plan

### PR DESCRIPTION
**Pull request by marnovandermaas**
This is an initial version of the I2C silicon validation test plan. It takes inspiration from the counterpart in key manager: https://github.com/lowRISC/opentitan/pull/19521

I tried to create enough tests so that all of I2C's features are covered, but it would probably be good to get the input of more people here. Alternatively, we can merge this as a base version and iterate over it with future PRs.

### CODE REVIEW

**Comment by marnovandermaas** [comment]
Note: I will be away next week, so if any changes need to be handled in that time someone else will need to pick it up.

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+              The test checks whether the payload is sent at the appropriate baud-rate
+            - For targets, the test sends data at the appropriate baud-rate and checks whether an ack is received.
+
+            Verify all instances of I2C in the chip.
+
+            Notes for silicon targets:
+            - The testbench in this case will be a device that physically toggles/inspects the appropriate SCL and SDA lines.
+            '''
+      features: ["I2C.MODE.HOST", "I2C.MODE.TARGET", "I2C.SPEED.STANDARD", "I2C.SPEED.FAST", "I2C.SPEED.FASTPLUS"]
+      stage: V2
```
Comment (moidx) [inline hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:71]: nit: `stage: V3` to avoid lowering the current regression completion results. Same for `chip_sw_i2c_override`.

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson [hunk 2]
```diff
@@ -0,0 +1,94 @@
+
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+{
+  name: chip_kmac
+  testpoints: [
```
Comment (moidx) [inline hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:7]: Should we add an explicit check for clock stretching to one of the tests?

**Review by moidx — APPROVED** [review]
Thanks @marnovandermaas! I added a couple of small comments. I think it is ok to submit this PR and then work on follow ups if needed. Thanks!

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            - SW validates the reception of tx empty and trans complete interrupts.
+            - Verify the virtual / true open drain capability.
+
+            Verify all instances of I2C in the chip.
+            '''
+      stage: V2
+      tests: ["chip_sw_i2c_device_tx_rx"]
+    }
+  ]
+}
```
Comment (engdoreis) [inline hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:94]: New line missing.

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson (hunk 2 above)
Comment (engdoreis) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:7]: Clock stretching is important, but I also suggest adding a test for NAK detection.

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson [hunk 4]
```diff
@@ -38,9 +43,52 @@
...
+      si_stage: SV2
       tests: ["chip_sw_i2c_device_tx_rx"]
     }
+    {
+      name: chip_sw_i2c_speed
+      desc: '''Verify the transmission speed of data over the chip's I2C interfaces.
+
+            Run the following for standard, fast and fast plus speeds:
+            - Program the I2C to be in the appropriate mode (host/target).
+            - For hosts, the SW test writes a known payload over the chip's I2C host interface.
```
Comment (engdoreis) [inline hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:61]: I suggest a loopback test where the host writes X and reads X+1 or X>>1.
Should we describe here how the baud rate will be measured (Since i2c is synchronous, the fact the data matches doesn't mean the speed is correct)?

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson (hunk 3 above)
Comment (marnovandermaas) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson]: Thanks for noticing this! This should now be fixed.

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson (hunk 2 above)
Comment (marnovandermaas) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:7]: Thanks for these suggestions, I tried to avoid specifying the whole I2C protocol in the feature list, but I agree that clock stretching and nack detection are important corner cases to test. So I have added them to the feature list and added two tests for them. Let me know what you think of them.

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson (hunk 4 above)
Comment (marnovandermaas) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:61]: Thanks for this suggestion. I've made an initial attempt at this, but happy to adjust it based on further feedback.

**Comment by marnovandermaas** [comment]
CHANGE AUTHORIZED: hw/ip/i2c/data/i2c.hjson

**Comment by marnovandermaas** [comment]
Since this PR is quite time sensitive, I propose that we merge this PR once CI passes and make any changes in follow-up PRs. Can another committer please authorize these changes:

CHANGE AUTHORIZED: hw/ip/i2c/data/i2c.hjson

File: hw/ip/i2c/data/i2c.hjson [hunk 5]
```diff
@@ -105,6 +105,63 @@
...
+            - During a write.
+            - During a read.
+            '''
+    },
+    { name: "I2C.PROTOCOL.NACK",
+      desc: '''
+            Whenever a byte is sent, it must be accompanied by an acknowledgement (ack), unless NAKOK is high.
+            When no ack is received, this is a nack.
+            '''
+    },
```
Comment (crteja) [inline hw/ip/i2c/data/i2c.hjson:163]: It would be good to add RSTART also as we have observed an issue with it.

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            For all I2C host mode instances:
+            - Program the I2C to be in host mode.
+            - Do a read and make target stretch after reading an address.
+            - Do a write and make the target stretch during that write.
+            - Do a read and make the target stretch during that read.
+
+            Notes for silicon targets:
+            - The testbench in this case will be a device that physically toggles/inspects the appropriate SCL and SDA lines.
+            '''
+      features: ["I2C.MODE.HOST", "I2C.PROTOCOL.CLOCKSTRETCHING"]
```
Comment (crteja) [inline hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:110]: [Clock stretching](https://opentitan.org/book/hw/ip/i2c/doc/theory_of_operation.html#clock-stretching) is supported in target mode as well and it would be good to have a separate test case for it.

Comment (marnovandermaas) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:110]: I specified this host test because during SiVal testing the testbench has control over the target and can do the clockstretching. How do you force the OpenTitan target to do clock stretching, even when the test bench only has control over the host?

File: hw/ip/i2c/data/i2c.hjson (hunk 5 above)
Comment (marnovandermaas) [reply hw/ip/i2c/data/i2c.hjson:163]: Thanks added repeated start as a feature and a test for it.

File: hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson (hunk 6 above)
Comment (crteja) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:110]: Agreed but I2C supports clock stretching in target mode as well (and is allowed in I2C protocol). For instance, when ACQ FIFO is full and Ibex is busy with other higher priority interrupts and delays reading from the FIFO, OT I2C target holds SCL low until ACQ FIFO has space for additional bytes. This scenario can be triggered in SiVal.

`chip_sw_i2c_clockstretching` only focuses on OT I2C host mode. As each I2C IP instance in OT is used either as a Host or a Target, , there should be a separate target mode test that covers clock stretching scenarios mentioned in https://opentitan.org/book/hw/ip/i2c/doc/theory_of_operation.html#clock-stretching. 
There is a DV test that covers some of these scenarios https://opentitan.org/book/hw/ip/i2c/data/i2c_testplan.html#target_clock_stretch

Comment (marnovandermaas) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:117]: Thank Ravi! I've made an initial attempt at this. Happy to change it though.

Comment (crteja) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:117]: Thanks Marno. This looks good for basic clock stretch in target mode. Do we need to exhaustively cover all cases of clock stretch in SiVal ?

Comment (johngt) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:117]: @crteja / @marnovandermaas 
For SiVal Phase-II we should cover base cases.
For SiVal Phase-III this is where we should cover exhaustive testing

Comment (crteja) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:117]: @marnovandermaas , would it make sense to have a separate SV3 test that covers all cases of clock stretching and mark this test as SV2 ?

Comment (marnovandermaas) [reply hw/top_earlgrey/data/ip/chip_i2c_testplan.hjson:117]: Ok, I've marked this as SV2 again.

**Review by rswarbrick — COMMENTED** [review]
CHANGE AUTHORIZED: hw/ip/i2c/data/i2c.hjson

**Comment by rswarbrick** [comment]
CHANGE AUTHORIZED: hw/ip/i2c/data/i2c.hjson
